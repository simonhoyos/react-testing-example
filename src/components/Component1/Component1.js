import React from 'react';

export function Component1() {
  return <h2>Componente 1</h2>;
}
