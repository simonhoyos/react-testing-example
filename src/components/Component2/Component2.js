import React from 'react';

export function Component2() {
  return <h2>Componente 2</h2>;
}
