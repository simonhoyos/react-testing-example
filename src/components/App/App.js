import React, { Component } from 'react';
import { Component1 } from '../Component1/Component1';
import { Component2 } from '../Component2/Component2';
import './App.css';

class App extends Component {
  state = {
    showFirst: true,
  };

  componentDidMount() {
    this.props.apiRequest();
  }

  handleChange = e => {
    const showFirst = e.currentTarget.checked;

    this.setState({ showFirst });
  }

  render() {
    return (
      <div className="App">
        <label htmlFor="check">Show First Component</label>
        <input
          type="checkbox"
          onChange={this.handleChange}
          id="check"
          checked={this.state.showFirst}
          data-testid="check"
        />
        {this.state.showFirst ?
          <Component1 /> :
          <Component2 />
        }
      </div>
    );
  }
}

export default App;
