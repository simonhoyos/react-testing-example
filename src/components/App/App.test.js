import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from './App';
import { Component1 } from '../Component1/Component1';
import { Component2 } from '../Component2/Component2';
import { ListItem } from '...'

describe('<App />', () => {
  it('should match snapshot', () => {
    const apiRequest = jest.fn();
    const wrapper = shallow(<App apiRequest={apiRequest} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('should render Component1 by default', () => {
    const apiRequest = jest.fn();
    const wrapper = shallow(<App apiRequest={apiRequest} />);

    // depende de la implementación
    expect(wrapper.state().showFirst).toBe(true);
    // independiente a la implementación
    expect(wrapper.find(Component1)).toHaveLength(1);
    expect(wrapper.find(Component2)).toHaveLength(0);
  });

  it('should render Component2 when checkbox changes', () => {
    const apiRequest = jest.fn();
    const wrapper = shallow(<App apiRequest={apiRequest} />);

    const checkbox = wrapper.find('[data-testid="check"]');
    const eventMock = {
      currentTarget: {
        checked: false,
      }
    }
    checkbox.simulate('change', eventMock);

    expect(wrapper.state().showFirst).toBe(false);

    expect(wrapper.find(Component1)).toHaveLength(0);
    expect(wrapper.find(Component2)).toHaveLength(1);
  });

  it('should render component2 when state showFirst is false', () => {
    const apiRequest = jest.fn();
    const wrapper = shallow(<App apiRequest={apiRequest} />);

    wrapper.setState({ showFirst: false });

    expect(wrapper.find(Component1)).toHaveLength(0);
    expect(wrapper.find(Component2)).toHaveLength(1);
  });

  // it('should get the data from API', () => {
  //   const dataMock = [
  //     {
  //       title: 'El Padrino',
  //       overview: 'lorem ipmsum',
  //     },
  //     {
  //       title: 'El Padrino',
  //       overview: 'lorem ipmsum',
  //     },
  //     {
  //       title: 'El Padrino',
  //       overview: 'lorem ipmsum',
  //     },
  //   ];
  //   const apiRequest = jest.fn();
  //   const wrapper = mount(<App apiRequest={apiRequest} />);

  //   App.prototype.componentDidMount();

  //   expect(apiRequest).toHaveBeenCalledTimes(1);

  //   wrapper.setState({ data: dataMock });

  //   expect(wrapper.find(ListItem)).toHaveLength(3);
  // });

  it('renders without crashing', () => {
    const apiRequest = jest.fn();
    const div = document.createElement('div');
    ReactDOM.render(<App apiRequest={apiRequest} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});
