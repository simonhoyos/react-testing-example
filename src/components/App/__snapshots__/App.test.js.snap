// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<div
  className="App"
>
  <label
    htmlFor="check"
  >
    Show First Component
  </label>
  <input
    checked={true}
    data-testid="check"
    id="check"
    onChange={[Function]}
    type="checkbox"
  />
  <Component1 />
</div>
`;
